gitpod-io/gitpod
# [ws-manager] gRPC rate limits and metrics

This PR introduces rate limits on the unary methods of ws-manager's gRPC interface. It also introduces Prometheus metrics for calls to that interface.

Many of ws-manager's methods directly hit the Kubernetes master, i.e. load on ws-manager is load on the Kubernetes data plane. In the past we've seen that broken clients (e.g. too many `ControlPort` calls - see #2401) can have a detrimental impact on the overall system performance. This this change, we can monitor such calls using metrics, and avoid such situations through rate limiting.

### How to test
Note: This branch carries a change to the preview environment that enables (far too severe) rate limiting for `StartWorkspace` and `ControlPort` - the former dropping calls, the latter waiting until there's a free slot. We'll remove https://github.com/gitpod-io/gitpod/commit/196add6d80c81b84103fea838c215b1fee8bc44d prior to merging.

1. start many workspaces on this branch quickly - some workspace starts should fail with `RESOURCE_EXHAUSTED`
2. in the running workspace, with the port view open, quickly start services on different ports, e.g.
    ```
    $ curl lama.sh | LAMA_PORT=9090 sh
    Ctrl+C
    $ curl lama.sh | LAMA_PORT=9091 sh
    ```
    You should see how the port detection slows down (because of an unfortunate coupling of concerns in supervisor) due to the now slow `ControlPort` call.
3. have a look at the ws-manager metrics, you should see the following counter:
    - `gitpod_ws_manager_grpc_call_total`
    - `gitpod_ws_manager_grpc_error_total`
